ENH: Add support for structured arrays and dtypes in assertion comparison functions

Micky774 opened the pull request:
## Related Issues/PRs
Fixes #21539

## Changes
Adds support for recursively traversing structured arrays for in-depth comparison.

## Comments

Micky774 committed:
Initial work towards structured equality assertion

Micky774 committed:
Merge branch 'main' into structured_equality

Also included minor refactor of recursive path

Micky774 committed:
Neaten

Micky774 committed:
Neaten

Micky774 committed:
Significant simplification of structured check and test corrections

- Test corrected to check against scalars w/ structured dtype, not
single-element ndarrays

- Refactored main check to go from recursive to iterative

Micky774 committed:
Flattened recursive helper to iterative, fixed tests

Micky774 committed:
Fixed remaining tests

Micky774 committed:
Linting

Micky774 committed:
Removed old variable

Micky774 committed:
Hopefully final linting change

Micky774 committed:
Reverted accidentally altered conditions for masked arrays

seberg commented on numpy/testing/_private/utils.py:
```diff
@@ -710,12 +710,93 @@ def assert_array_compare(comparison, x, y, err_msg='', verbose=True, header='',
+                if _x.dtype.names is None:
+                    msg = check_flagged_comparison(_x, _y, **kwargs)
+                    if msg:
+                        return msg
```
I am wondering if rather than flattening here to a lits, we should use a recursive generator:
```
def get_views(arr, field_name=()):
    if not arr.dtype.names:
        yield arr, field_name
    else:
        for name in arr.dtype.names:
            yield from get_views(arr[name], field_name+(name,))
```

Zac-HD commented:
Wrote a Hypothesis test for this:

```python
import numpy as np
from hypothesis import given
import hypothesis.extra.numpy as npst

@given(
    npst.arrays(
        dtype=npst.nested_dtypes(),  # includes subarrays
        shape=npst.array_shapes(max_side=3),
    )
)
def test_all_arrays_are_equal_to_themselves(arr):
    np.testing.assert_array_equal(arr, arr)
```

Micky774 committed:
Simplified implementation

Micky774 committed:
Added hypothesis test

Micky774 committed:
Simplified test further

Micky774 committed:
Reverted format change for diff

Micky774 committed:
Merge branch 'main' into structured_equality

Micky774 committed:
Simplify assertion error raising

Micky774 committed:
inlined function

Micky774 commented:
@seberg I simplified the implementation a great deal thanks to your suggestion. Also incorporated @Zac-HD's awesome hypothesis test. I think it should be ready for another look.

seberg commented:
Thanks, some test still seems to be failing though.  Although it is a segfault, so it probably just found anothr bug :|

seberg commented on numpy/lib/tests/test_io.py:
```diff
@@ -2728,4 +2728,7 @@ def test_load_refcount():
+        assert_equal(x, np.array(((0, 1), (2, 3)), dtype=dt))
+        # assert_equal uses a recursive function which
+        # makes this context unhappy.
+        break_cycles()
```
Can you reorganize it to move the `assert_equal` out of the `with` statement?  What we care about is loadtxt, not the assertion function.

Micky774 replied:
Done.

Micky774 committed:
Fixed bug

Micky774 committed:
Updated test

Micky774 committed:
Newline for linter

Micky774 committed:
Simplified variable passing

Micky774 commented:
> Thanks, some test still seems to be failing though. Although it is a segfault, so it probably just found anothr bug :|

@seberg should be resolved now

Micky774 committed:
Removed trivial if statement

Micky774 committed:
Merge branch 'main' into structured_equality

Thijss commented:
@Micky774 Any updates on this PR?

Micky774 commented:
> @Micky774 Any updates on this PR?

IIRC the material change is ready, but this PR is quite old so it'll need to be reconciled with `main` and any API changes since then -- which hopefully isn't more than trivial changes.

@seberg Is this PR still useful and worth pursuing again, or have there been changes in NumPy that would warrant a different approach (or render it moot)?

seberg commented:
No, I don't think things dramatically changed.  It might be nice to make structured comparisons a ufunc, then we could have `equal_nans` or so as a ufunc, which could be rather helpful and would solve the issue at hand.

But, the point of this PR was always to fix an aparent need, rather than to find an ideal solution...
